KEP-3329 Add KEP for Retriable and non-retriable Pod failures for Jobs (#3374)
kubernetes/enhancements: keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

- [TensorFlow Training (TFJob)](https://www.kubeflow.org/docs/components/training/tftraining/)
- [Argo workflow](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml) ([example](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml))

Additionally, some pod failures are not linked with the container execution,
but rather with the internal kubernetes cluster management (see:
[Scheduling, Preemption and Eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/)).
Such pod failures should be recongnized as infrastructure failures and it
should be possible to ignore them from the counter towards `backoffLimit`.

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals

- Extension of Job API with user-friendly syntax to terminate jobs based on the

>   end state of the failed pod/container.

Reviewer: Even before the other proposal (#3339), I wondered do we want to go all the way down to containers. Why not just focus on pod level failures, for starters?

Author: The idea of using container exit codes is to satisfy the requirement raised in issues: https://github.com/kubernetes/kubernetes/issues/17244 or https://github.com/kubernetes/kubernetes/issues/31147.

Some frameworks like TF or Argo support retries based on container exit codes so it seems reasonable to support it natively. 

Also, having two retry strategies: based on exit codes and pod termination reasons can hopefully lead to a more flexible JobSpec API.

addition of a new reason will require an API review. The constants will allow
users of the package to reduce the risk of typing mistakes.

#### Garbage collected pods

The Pod status (which includes the `conditions` field and the container exit
codes) could be lost if the failed pod is garbage collected.

Losing Pod's status before it is interpreted by Job Controller can be prevented
by using the feature of [job tracking with finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/).

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### New PodConditions

The already existing `status.conditions` field in Pod will be used by kubernetes
components (such as: kube-scheduler, kubelet pod-gc) to append a dedicated
condition when they send the delete operation.

During the implementation process we are going to review the places where the

> pod delete requests are issued to modify the code to also append a meaningful

Author: added a note

	// This is an action which might be taken on pod failure - the pod will be
	// restarted and the counter for .backoffLimit will not be incremented.
	BackoffActionIgnore BackoffPolicyAction = "Ignore"
)

type BackoffPolicyOnExitCodeInterval struct {
	MinInclusive int
	MaxInclusive int
}

type BackoffPolicyOnExitCodeRequirement struct {
	// Restricts the check for exit codes to only apply to the container with
	// the specified name. When empty the rule applies to all containers
	// +optional
	ContainerName *string

	// Specifies the set of intervals. The requirement is satisfied when the
	// actual exit code is in any of the intervals
	// +optional
	InRange *[]BackoffPolicyOnExitCodeInterval

	// Specifies the set of intervals. The requirement is satisfied when the
	// actual exit code is not in any of the intervals
	// +optional
	NotInRange *[]BackoffPolicyOnExitCodeInterval

	// Specifies the set of exit codes. The requirement is satisfied when the
	// actual exit code is in the set
	// +optional
	In *[]int

	// Specifies the set of exit codes. The requirement is satisfied when the
	// actual exit code is not in the set
	// +optional

> 	NotIn *[]int

Reviewer: Nit: I don't think we need pointers to arrays, it's sufficient for array to be optional.

Author: done

}

type BackoffPolicyOnPodConditionOperator string

const (
	BackoffPolicyOnPodConditionOperatorIn    BackoffPolicyOnPodConditionOperator = "In"
	BackoffPolicyOnPodConditionOperatorNotIn BackoffPolicyOnPodConditionOperator = "NotIn"
)

type BackoffPolicyOnPodConditionsRequirement struct {

	// Represents the name of the field in the pod condition

> 	Key string

Reviewer 2: We should always be matching with a Type.

Reviewer 2: Just by looking at how the API is getting complicated, I really don't think we should be looking at reasons. See the documentation:

```
	// Unique, one-word, CamelCase reason for the condition's last transition.
	// +optional
	Reason string
```

And all the possible reasons are not really documented. So I vote we ignore them in this API.

Maybe we can do something like this:

```golang
type BackoffPolicyOnPodConditionsRequirement struct {
  // Could be AnyOf, AllOf
  Operator BackoffPolicyOnPodConditionOperator
  Conditions []PodConditionMatcher
}

type PodConditionMatcher {
  Type PodConditionType
  // optional, true by default
  Status *PodConditionStatus
```

And if we later decide that comparing reasons is necessary, we can expand the API by adding `Reasons []string`

Reviewer 3: > We should always be matching with a Type.

I agree with this. Without a specific condition type, the meaning of the status field is undefined.

Reviewer 3: If the intent is to retry only on infrastructure errors, that is not a well-defined set of conditions.

Currently, that includes things like Evicted, Orphaned, Preempted, NodeShutdown, but that is not necessarily a bounded set, and could grow in the future (just as NodeShutdown was added a couple releases ago).

We don't want to make job authors responsible for discovering/enumerating all the various condition types related to infrastructure pod failures. If they do, that means they are likely to miss rarely encountered conditions, and if we add new infrastructure conditions in the future, previously "correct" job specs would become incorrect.

A couple things could help here:
1. make the retry policy only go in the `Terminate` direction... it should opt out of retrying for specific conditions or exit codes (e.g. "skip retrying on x,y,z")... that fails in a safer direction if a rare condition is accidentally missed
2. it would be more user-friendly if a single condition identified when a pod terminated "normally" rather than because of external reasons (infrastructure of pod deletion)

Reviewer 2: Things we seem to agree:
- We need an explicit field to match the PodCondition.Type
- Since conditions already have an enumeration problem, we probably don't want to make it worse by paying attention to the "reason", which is more free form than type.

> A couple things could help here:

I think either 1 or 2 are sufficient to prevent authors from having to discover infrastructure errors.
If we go with 1, a provider can only express "retry up to `backoffLimit` infrastructure errors, but terminate on first user error". But it can't express "retry indefinitely on infrastructure errors, but retry up to `backoffLimit` user errors"
Doing just 2 allows to express both.

- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP extends Kubernetes to determine if a pod failure of a job is retriable

> or not. Pod failures determined as non-retriable will result in early job termination.

Reviewer 2: Quickly mention the retriable infrastructure errors as well

that the root cause of a failure is in the executable and the
job is destined to fail regardless of the number of retries. However, since
the large workloads are often scheduled to run over night or over the
weekend, there is no human assistance to terminate such a job early.

The need for solving the problem has been emphasized by the kubernetes
community in the issues, see: [#17244](https://github.com/kubernetes/kubernetes/issues/17244) and [#31147](https://github.com/kubernetes/kubernetes/issues/31147).

Some third-party frameworks have implemented retry policies for their pods:
- [TensorFlow Training (TFJob)](https://www.kubeflow.org/docs/components/training/tftraining/)
- [Argo workflow](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml) ([example](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml))

Additionally, some pod failures are not linked with the container execution,
but rather with the internal kubernetes cluster management (see:
[Scheduling, Preemption and Eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/)).

> Such pod failures should be recongnized as infrastructure failures and should
> not increment the counter for `backoffLimit`.

Reviewer 2: ```suggestion
Such pod failures should be recongnized as infrastructure failures and it should be possible to ignore them from the counter towards `backoffLimit`.
```

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals

- Extension of Job API with user-friendly syntax to terminate jobs based on:
  * container exit codes
  * `status.reason` field of a failed pod.
- Extensible API which will allow to implement support for other indicators
  of non-retriable jobs in the future, for example based on the termination
  logs (see: [Determine the Reason for Pod Failure](https://kubernetes.io/docs/tasks/debug/debug-application/determine-reason-pod-failure/)).

<!--
TODO: consider adding the point:
- Review of the places where the `DeleteOptions` structure is used to delete
  pods and specification of the `reason` field in these contexts.
-->
<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

### Non-Goals

> - Implementation of other indicators of non-retirable jobs such as termination logs.

Reviewer 2: ```suggestion
- Implementation of other indicators of non-retriable jobs such as termination logs.
```

> - Modificaiton of the semantics for job termination. In particular, allowing for

Reviewer 2: ```suggestion
- Modification of the semantics for job termination. In particular, allowing for
```

  all indexes of an indexed-job to execute when only one or a few indexes fail

>   [#109712](https://github.com/kubernetes/kubernetes/issues/109712).

Reviewer 2: Also mention that other APIs such as Deployment, StatefulSets are not covered.

run at night or over weekend without any human monitoring. In order to account
for random infrastructure failures we define `.backoffLimit: 6` for the job.
However, a signifficant portion of the failures happen due to bugs in code.
Moreover, the failures may happen late during the program execution time. In
such case, restarting such a pod results in wasting a lot of computational time.

We would like to be able to automatically detect and terminate the jobs which are
failing due to the bugs in code of the executable, so that the computation resources
can be saved.

Occasionally, our executable fails, but it can be safely restarted with a good
chance of succeeding the next time. In such known retriable situations our
executable exits with a dedicated exit code in the 40-50 range. All remaining
exit codes indicate a software bug and should result in an early job termination.

The following Job configuration could be a good starting point to satisfy
my needs:

```yaml
apiVersion: v1
kind: Job
spec:
  template:
    spec:
      containers:
      - name: job-container
        image: job-image
        command: ["./program"]
  backoffLimit: 3
  backoffPolicy:
    rules:

>     - containerExitCodeNotIn: 40-50

Reviewer 2: I think this should be something like:

```yaml
containerExitCode:
  notIn: 40-50
```

or

```yaml
containerExitCode:
  value: 40-50
  operator: NotIn
```

Author: I was thinking about it, but I feel that another level of nesting is over-complicating the API.

Reviewer 4: Is it really a good idea to specify two numbers in a single string `40-59`? This feels like a YAML parsing trap, and creates trouble for automation (e.g. I can't easily `jq '.spec.backoffPolicy.rules[0].containerExitCodeNotIn.max = 59`).

Author: Good points, I will define ranges (min, max) and extract the "in" and "notIn" to separate levels.
See the proposal below (I will update the KEP accordingly):

```
  backoffPolicy:
    rules:
    - action: Terminate
      containerName: main-job-container
      onContainerExitCode:
        inRange:
        - min: 1
          max: 12
        - min: 100
          max: 127
    - action: Ignore
      onPodStatusReason:
        in:
        - Preemption
        - NodePressureEviction
```

Reviewer 4: So each exit code filter is an `IntOrIntRange`?

Reviewer 4: Also note the possible off-by-one ambiguity. Does `{min: 1, max: 12}` include 12 or not?

Author: @SOF3 how about:
- introducing "inRanges" along with "in" (to avoid "IntOrIntRange")
- rename the min, max -> minInclusive and maxInclusive. 

Then, the example could be formulated as:
```
  backoffPolicy:
    rules:
    - action: Terminate
      containerName: main-job-container
      onContainerExitCode:
        inRanges:
        - minInclusive: 1
          maxInclusive: 12
        - minInclusive: 100
          maxInclusive: 127
    - action: Ignore
      onPodStatusReason:
        in:
        - Preemption
        - NodePressureEviction
```

Author: Update, in the latest API proposal this can be expressed as follows (also notice the change too lookup for PodCondition rather than Pod StatusReason):

```yaml
 backoffPolicy:
   rules:
   - action: Terminate
     requirements:
     - onExitCodes:
         containerName: main-job-container
         inRange:
         - minInclusive: 1
           maxInclusive: 10
         - minInclusive: 100
           maxInclusive: 127
   - action: Ignore
     requirements:
     - onPodCondition:
       - key: Status
         operator: In
         values:
         - True
       - key: Type
         operator: In
         values:
         - Preempted
         - Evicted
```

  backoffPolicy:
    rules:
    - podStatusReasonIn:
      - Preemption
      - NodePressureEviction
      action: Ignore
```

Note that, in this case the user supplies a list of pod's `status.reason` values
which are interpreted as infrastructure failures. This approach is likely to
require an iterative process to review and extend of the list. In order to
avoid the process some users may prefer to interpret any non-empty `status.reason`
as an infrastructure failure this can be achieved with the following syntax for
`backoffPolicy`:

```yaml
  backoffPolicy:
    rules:
    - podStatusReasonNotIn:
      - ''
      action: Ignore
```

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

#### Risk 1

The list of available options for `reason` field will be evolving with new
values being added and potentially some values becoming obsolete. This can make
it difficult to maintain a valid list of reasons enumerated in the
Job configuration.

> First, the users of the Job configuration will be able to mitigate it with the

Reviewer 2: A mitigation should be on us to do, not on the user.

We can do 2 things:

- Document the existing reasons.
- Move all the reasons values to the `k8s.io/api/core/v1`. This would reduce typing mistakes and force an API reviewer to approve every new reason.

API which will allow them to specify that any failure of pod with non-empty
`status.reason` can be ignored. This will eliminate the need of listing all the
values. Second, if a user decides to list of all `reasons` to be ignored, the
consequence of a new reason added in a new version of kubernetes is limited -
the failed pod will increment the counter towards the `backoffLimit` which is
the current behaviour.

> #### Risk 2

Reviewer 2: ```suggestion
#### Garbage collected pods
```

kube-scheduler, kubelet pod-gc) when they send the delete operation. During
the implementation process, we are going to review the places were the
`DeleteOptions` is used to delete pods and add the `reason` field based on the
invocation context.

Notably, the `reason` field might be useful to be supplied when deleting other
types of kubernetes resources, but supplying the field in other contexts is
beyond this KEP.

In the case of Pods, kube-apiserver will store the received `reason` in the
already existing `reason` field in the `PodStatus`:
```
// PodStatus represents information about the status of a pod. Status may trail the actual
// state of a system.
type PodStatus struct {
  ...
	// A brief CamelCase message indicating details about why the pod is in this state. e.g. 'Evicted'
	// +optional
	Reason string
  ...
}
```

### JobSpec API

We extend the Job API in order to allow to apply different actions depending
on the conditions associated with the pod failure.

```golang
// BackoffPolicyAction specifies how a Pod failure is handled.
type BackoffPolicyAction string

const (

	// This is an action which might be taken on pod failure - mark the

> 	// pod's job as Failed.

Reviewer 2: and terminate all the running pods.

	TerminateBackoffAction BackoffPolicyAction = "Terminate"

	// This is an action which might be taken on pod failure - the pod will be
	// restarted and the counter for .backoffLimit will not be incremented.
	IgnoreBackoffAction BackoffPolicyAction = "Ignore"

)

// BackoffPolicyRule specifies a backoff policy rule, which comprises an action
// taken on a pod failure and the condition which needs to be satisfied to take
// the action.
type BackoffPolicyRule struct {

	// Specifies the action taken on a pod failure when the condition is satisfied.
	Action BackoffPolicyAction

	// Restricts the check for exit codes to only apply to the container with
	// the specified names.
	// +optional
	ContainerName *string

	// Specifies the rule condition by the comma-separated list of exit codes or

> 	// ranges of exit codes. The condition is satisfied if the actual container's

Reviewer 2: You can say things like:

When this field is in use, field X needs to be empty.

	// specify the set of actions and conditions which need to be
	// satisfied to take the associated action.
	// If empty, the default behaviour applies - the counter of pod failed is
	// incremented and it is checked against the backoffLimit
	// +optional
	BackoffPolicy *BackoffPolicy
  ...
```

Additionally, we validate the following constraints for each instance of `BackoffPolicyRule`:
- exactly one of the following fields is used:`ContainerExitCodeIn`,
 `ContainerExitCodeNotIn`, `PodStatusReasonIn`, `PodStatusReasonNotIn`.
- the `ContainerName` is only combined with the fields `ContainerExitCodeIn` or
 `ContainerExitCodeNotIn`.

Here is an example Job configuration which uses this API:

```yaml
apiVersion: v1
kind: Job
spec:
  template:
    spec:
      containers:
      - name: main-job-container
        image: job-image
        command: ["./program"]
      - name: monitoring-job-container
        image: job-monitoring
        command: ["./monitoring"]
  backoffLimit: 3
  backoffPolicy:
    rules:
    - containerExitCodeIn: 1-12,100-127
      containerName: main-job-container
      action: Terminate
    - podStatusReasonNotIn:
      - ''
      action: Ignore
```

> <!--

Reviewer 2: Remove this

This is important to clarify: What does this mean?

Would the rest of the pods be terminated? Would they allow to run? The simplest approach is to terminate all the pods, just like we do when the backoffLimit is reached. And we can introduce other options (and some specific to indexed jobs) in a follow up KEP next release.
 -->

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### Test Plan

<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md

> -->

Reviewer 2: You can say that the current job-controller has adequate testing for backoff. The SIG approver will say if that makes sense or not.

> [ ] I/we understand the owners of the involved components may require updates to

Reviewer 2: Mark this

existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

##### Unit tests

<!--
In principle every added code should have complete unit test coverage, so providing
the exact set of tests will not bring additional value.
However, if complete unit test coverage is not possible, explain the reason of it
together with explanation why this is acceptable.
-->

<!--
Additionally, for Alpha try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>
The data can be easily read from:
https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

> - `<package>`: `<date>` - `<test coverage>`

Reviewer 2: Please complete this.

for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [x] Feature gate (also fill in values in `kep.yaml`)

>   - Feature gate name: RetriableAndNonRetriableFailures

Reviewer 2: it should have Job in the name. But it could be something like: JobBackoffPolicy

  - Components depending on the feature gate:

>     - kube-controller-manager

Reviewer 2: kube-apiserver and kube-scheduler too

Author: Actually, I'm thinking about what exactly needs to be hidden behind the feature gate. The job controller handling of failures - clearly, thus the kube-controller-manager. 
However, we could just ignore the reasons at the kube-apiserver or prevent even sending the reasons - then kube-scheduler as well.

Author: I' m going to assume we hide all the parts of the solution behind the feature gate.

Reviewer 2: > However, we could just ignore the reasons at the kube-apiserver or prevent even sending the reasons - then kube-scheduler as well.

Yes, and that needs to be done when the feature gate is disabled.

previous answers based on experience in the field.
-->

###### Will enabling / using this feature result in any new API calls?

No.

<!--
Describe them, providing:
  - API call type (e.g. PATCH pods)
  - estimated throughput
  - originating component(s) (e.g. Kubelet, Feature-X-controller)
Focusing mostly on:
  - components listing and/or watching resources they didn't before
  - API calls that may be triggered by changes of some Kubernetes resources
    (e.g. update of object X triggers new updates of object Y)
  - periodic API calls to reconcile state (e.g. periodic fetching state,
    heartbeats, leader election, etc.)
-->

###### Will enabling / using this feature result in introducing new API types?

No.
<!--
Describe them, providing:
  - API type
  - Supported number of objects per cluster
  - Supported number of objects per namespace (for namespace-scoped objects)
-->

###### Will enabling / using this feature result in any new calls to the cloud provider?

No.
<!--
Describe them, providing:
  - Which API(s):
  - Estimated increase:
-->

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

> Yes. The `DeleteOptions` structure is extended by the `reason` field.

Reviewer 2: `DeleteOptions` is not stored, so there is nothing to increase.

The rest of the things are opt-in, so we wouldn't increase the size of existing objects.

run at night or over weekend without any human monitoring. In order to account
for random infrastructure failures we define `.backoffLimit: 6` for the job.
However, a signifficant portion of the failures happen due to bugs in code.
Moreover, the failures may happen late during the program execution time. In
such case, restarting such a pod results in wasting a lot of computational time.

We would like to be able to automatically detect and terminate the jobs which are
failing due to the bugs in code of the executable, so that the computation resources
can be saved.

Occasionally, our executable fails, but it can be safely restarted with a good
chance of succeeding the next time. In such known retriable situations our
executable exits with a dedicated exit code in the 40-50 range. All remaining
exit codes indicate a software bug and should result in an early job termination.

The following Job configuration could be a good starting point to satisfy
my needs:

```yaml
apiVersion: v1
kind: Job
spec:
  template:
    spec:
      containers:
      - name: job-container
        image: job-image
        command: ["./program"]

>   backoffLimit: 3

Author: updated

  all indexes of an indexed-job to execute when only one or a few indexes fail
  [#109712](https://github.com/kubernetes/kubernetes/issues/109712).
- Similar termination policies for other workload controllers such as Deployments
  or StatefulSets.

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

Extension of the Job API with a new field which allows to configure the set of
conditions which determine how a pod failure is handled. We extend the Job API
to support discrimination of pod failures based on container exit codes as well
as the pod's `status.reason` field.

<!--
The value of status.reason I found used by k8s currently is:
	// NodeUnreachablePodReason is the reason on a pod when its state cannot be confirmed as kubelet is unresponsive
	// on the node it is (was) running.
NodeUnreachablePodReason = "NodeLost" (defined on node.go)
-->
The pod's `status.reason` field is currently defined but used in few situations.
We extend its use to store the reason of a pod's failure when originated by
a kubernetes component (kube-scheduler, kubelet, pod-gc, etc.). To achieve this

> we extend the Pod delete API to pass the reason for pod termination. Such

Reviewer 5: I recommend not changing the delete api, since it's not reasonable to do this specially for specific apis (such as pods). Since that code is generic.

Instead just change these components to set a status condition before deleting. Since you have to change them anyway.

Reviewer 2: 1. We want to target the existing Pod.status.reason field, which kubelet already uses. And one of the kcm node controllers (an update right before calling Delete)
2. The alternative you propose is certainly an option, but it has the disadvantage of adding API calls. Preemption is already a slow process in kube-scheduler (one delete for each low priority pod that needs to be deleted). I would prefer not to add another API call. But we can certainly go this path if api-machinery strongly disagrees.

Reviewer 2: 3. I don't see why other resources could not benefit from having a reason in their status, but it's out-of-the scope of this KEP.

Reviewer 5: Sorry, I missed that you were not talking about a condition, but that doesn't change my answer.

pod's status.reason sounds like a field that should not exist if we'd gotten conditions right to begin with. We would not want to add a way to set that field on delete on other objects.

There's a small chance you can figure out a way to do this with mutating webhooks, in which case no one can really stop you, but I still recommend against doing this. (I also think it probably won't work but I'm not sure specifically what prevents it)

I don't think it is bad to do an extra API call to both update and delete. If you are worried about QPS, don't be, I think premption is a very high-priority item and we can get you more throughput to do it if it's needed.

Reviewer 5: BTW, I think it's a very bad idea to have a single field with multiple controllers (kubelet + controller manager + new use) that all want to set it. I think you should add a new field, or make a new condition. The api is intended to help controllers set *different parts* of the object, not to share responsibility for a single field, I'm not sure how that could work. Kubelet could clobber your write, for example.

Reviewer 2: > I think it's a very bad idea to have a single field with multiple controllers that all want to set it.

Fair, we can probably change the proposal to use PodConditions instead. We just have to change kubelet to also do the same (which means we need sig-node as participating-sig?)

> If you are worried about QPS, don't be, I think premption is a very high-priority item and we can get you more throughput to do it if it's needed.

Sounds good. We can put "tweaking the scheduler priority" that as a beta criteria based on scalability testing.

Reviewer 2: Discussed offline with @lavalamp:

If we are going the path of an update before a delete, then this is of no concern of sig-api-machinery and can be removed from participating sigs.

Let's add sig-node.

Reviewer: :+1: for using PodCondition

Author: The KEP uses now `pod.status.conditions` field instead of the `status.reason` field. Also, there is a comment of applying rules when we have multiple conditions:

Note that, a pod can have multiple Pod conditions with multiple types present. A rule is applied if any of the conditions matches the specified rule requirements. Also, it is only the first rule with matching requirements which is applied as the rules are evaluated in order.

failures can typically be associated with some cluster-management actions and
should be retried.

> We use the main job controller main loop to detect and categorize the pod failures

Author: updated

What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

#### Evolving reasons

The list of available options for `reason` field will be evolving with new
values being added and potentially some values becoming obsolete. This can make
it difficult to maintain a valid list of reasons enumerated in the
Job configuration.

First, the consequence of a new reason added in a new version of
kubernetes is limited - when a failed pod is not matched against any configured
rule in job's `spec.backoffPolicy` the default handling applies.

Second, In order to mitigate this risk we are going to document and define all
reasons as constants in the k8s.io/api/core/v1 package. As a result, every
addition of a new reason will require an API review. The constants will allow
users of the package to reduce the risk of typing mistakes.

#### Garbage collected pods

The Pod status (which includes the `reason` field and the container exit codes)
could be lost if the failed pod is garbage collected.

First, it should be rather

> rare so the overall consequence of unnecessary pod restarts will be limited.

Author: updated

	BackoffActionIgnore BackoffPolicyAction = "Ignore"

)

// BackoffPolicySource specifies the source of the failure
type BackoffPolicySource string

const (

	BackoffPolicySourceExitCode BackoffPolicySource = "ExitCode"

	BackoffPolicySourceStatusReason BackoffPolicySource = "StatusReason"

)

type BackoffPolicyRequirementOperator string

const (
	BackoffPolicyRequirementOperatorIn                      BackoffPolicyRequirementOperator = "In"
	BackoffPolicyRequirementOperatorNotIn                   BackoffPolicyRequirementOperator = "NotIn"
  BackoffPolicyRequirementOperatorInClosedRange           BackoffPolicyRequirementOperator = "InClosedRange"
  BackoffPolicyRequirementOperatorNotInClosedRange        BackoffPolicyRequirementOperator = "NotInClosedRange"
)

// BackoffPolicy describes how failed pods influence the backoffLimit.
type BackoffPolicy struct {
	// A list of backoff policy rules. The rules are evaluated in order.
	// Once a rule matches a Pod failure, the remaining of the rules are ignored.
	Rules []BackoffPolicyRule
}

type BackoffPolicyRule struct {
	// Specifies the action taken on a pod failure when the requirements are satisfied.
	Action BackoffPolicyAction

> 	// A list of backoff policy requirements. The requirements are ANDed.

Reviewer 2: ```suggestion
	// A list of backoff policy requirements that need to be satisfied to apply this rule. The requirements are ANDed.
```

	Requirements []BackoffPolicyRequirement
}

type BackoffPolicyRequirement struct {
	// Specifies the source of the failure. Valid sources are:
	// - ExitCode: The exit code from a container
	// - StatusReason: The reason for the Pod failed status.
	Source BackoffPolicySource

	// Restricts the check for exit codes to only apply to the container with
	// the specified name. When empty the rule applies to all containers
	// +optional
	ContainerName *string

	// Represents the source’s relationship to a set of values.

> 	// Valid Operators are In, NotIn, Gt and Lt.

Reviewer 2: Also InClosedRange

Author: I no longer enumerate all operators here in the comment - there a list of them above

	Operator BackoffPolicyRequirementOperator

	// An array of string values. If the operator is In or NotIn,
	// the values array must be non-empty. If the operator is Gt or Lt, the values
	// array must have a single element, which will be interpreted as an integer.
	Values []string
}

// JobSpec describes how the job execution will look like.
type JobSpec struct {
  ...
	// Specifies the policy of handling failed pods. In particular, it allows to
	// specify the set of actions and conditions which need to be
	// satisfied to take the associated action.
	// If empty, the default behaviour applies - the counter of pod failed is
	// incremented and it is checked against the backoffLimit
	// +optional

> 	BackoffPolicy *BackoffPolicy

Reviewer 2: Surfacing discussion from offline discussion:

Should we call this FailurePolicy or PodFailurePolicy instead?

useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

## Motivation

Running a large computational workload, comprising thousands of pods on
thousands of nodes requires usage of pod restart policies in order
to account for infrastructure failures.

Currently, kubernetes Job API offers a way to account for infrastructure
failures by setting `.backoffLimit > 0`. However, this mechanism intructs the
job controller to restart all failed pods - regardless of the root cause
of the failures. Thus, in some scenarios this leads to unnecessary
restarts of many pods, resulting in a waste of time and computational
resources. What makes the restarts more expensive is the fact that the
failures may be encountered late in the execution time of a program.

> Sometimes it can be determined from containers exit codes

Reviewer 6: Is it possible to achieve this at pod Level, it seems reusable. FYI I'm going to add a newfield `maxRestartTimes` to pod spec https://github.com/kubernetes/enhancements/pull/3339, it's still under discussion and waiting for review. However, if the idea accepted, the feature is somehow overlap with `backoffLimit`. This situation is similar to what this KEP wants to introduce.

Reviewer: I think your proposal will be a nice building block we can easily build on in the jobs.

Reviewer 2: when the PodSpec's `restartPolicy=Never`, it doesn't make sense to have an exitCode policy in the PodSpec. So in that case we will still need a policy in the JobSpec (this proposal).

In other words, the proposals can coexist.

Such pod failures should be recongnized as infrastructure failures and it
should be possible to ignore them from the counter towards `backoffLimit`.

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals

- Extension of Job API with user-friendly syntax to terminate jobs based on the
  end state of the failed pod/container.

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

### Non-Goals

- Implementation of other indicators of non-retriable jobs such as termination logs.
- Modification of the semantics for job termination. In particular, allowing for
  all indexes of an indexed-job to execute when only one or a few indexes fail
  [#109712](https://github.com/kubernetes/kubernetes/issues/109712).

> - Similar termination policies for other workload controllers such as Deployments

Author: We discuss in Notes the reasoning behind implementing this proposal at the job-level.

// container
type ExitCodeCondition struct {
	// specifies the set of internvals. The condition is satisfied then the
	// actual exit code value is in any of the intervals
	// +optional
	in      *[]ExitCodeInterval

	// specifies the set of internvals. The condition is satisfied then the
	// actual exit code value is not in any of the intervals
	// +optional
	notIn   *[]ExitCodeInterval

	// Restricts the check for exit codes to only apply to the container with
	// the specified name
	// +optional
	ContainerName *string
}

// PodStatusReasonCondition specifies the requirements for pod's status.reason
type PodStatusReasonCondition struct {

	// specifies the set of string values. The condition is satisfied when the
	// actual value of pod's status reason is in this set
	// +optional
	in *[]string

	// specifies the set of string values. The condition is satisfied when the
	// actual value of pod's status reason is not in this set
	// +optional
	notIn *[]string
}

// BackoffPolicyRule specifies a backoff policy rule, which comprises an action
// taken on a pod failure and the condition which needs to be satisfied to take
// the action.

> type BackoffPolicyRule struct {

Reviewer 7: I am afraid having new operator names like InClosedRange/NotInClosedRange will add some confusion, users may try to use them in labelSelector for example. Are they really necessary?

Also, it seems we have some canonical definitions here which include gt and lt (it doesn't cover the range case though): https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apimachinery/pkg/selection/operator.go#L23

> type BackoffPolicyRule struct {

Author: They are not really necessary. However, they are useful to make syntax less verbose in common cases. 

Say, you want to terminate the job if the container exit code is in range [1,12]. 

Using the lt, gt operators only:
`      
    requirements:
      - source: ExitCode
        containerName: main-job-container
        operator: gt
        Values:
        - 0
      - source: ExitCode
        containerName: main-job-container
        operator: lt
        Values:
        - 13
`
(which means also duplication of the containerName field) vs using the dedicated operator:
`
    requirements:
      - source: ExitCode
        containerName: main-job-container
        operator: InClosedRange
        Values:
        - 1
        - 12
`

> type BackoffPolicyRule struct {

Author: Update in the latest proposal I drop the idea of using operators for restricting exit codes. Reasons:
- they might be misleading anyway for users of label selectors as they don't use the key field
- their use is more verbose when you want to express a relatively simple condition: fail the job if the exit code for the "main-job-container" is in range [1,10] or [100, 127]. This can be achieved in the latest API:

```yaml
 backoffPolicy:
   rules:
   - action: Terminate
     requirements:
     - onExitCodes:
         containerName: main-container
         inRange:
         - minInclusive: 1
           maxInclusive: 10
         - minInclusive: 100
           maxInclusive: 127
```

> type BackoffPolicyRule struct {

Author: The API is updated - please refer to KEP for the current version. Some alternatives are discussed there too.

race-conditions between Kubelet and Job controller. For example, Kubelet
could restart a failed container before the Job controller decides to terminate
the corresponding job due to a rule using `onExitCodes`. On the other hand,
the unnecessary container restart may not be too much of an issue. We are going
to re-evaluate if we want to add support for `onExitCodes` combined with
`restartPolicy=OnFailure` in Beta.

#### Job-level vs. pod-level spec

We considered introduction of this feature at pod-level allowing to account for
container restarts within a running pod. However, we consider handling of
job-level failures (for example caused by Preemption) as an intetral part of
this feature. Also, when `restartPolicy` is specified as `Never`, then the
failures are handled at job-level anyway. If we establish a problematic
interaction of this feature with another feature providing container restarts
at the pod-level we will consider additional validation of the JobSpec
configuration to avoid the situation.

#### JobSpec API alternatives

Alternative versions of the JobSpec API to define requirements on exit codes and
on pod end state have been proposed and discussed under the

> [#3374](https://github.com/kubernetes/enhancements/pull/3374) PR. The outcome

Reviewer 2: Link to the alternatives section instead.

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### New PodConditions

The already existing `status.conditions` field in Pod will be used by kubernetes
components (such as: kube-scheduler, kubelet pod-gc) to append a dedicated
condition when they send the delete operation. The appended condition will have
a new dedicated type, called `SystemTermination`.

The API call to append the condition will be issued as a pod status update call
before the Pod delete request. This way the Job controller will already see the
condition when handling a failed pod.

During the implementation process we are going to review the places where the
pod delete requests are issued to modify the code to also append a meaningful
condition with dedicated `Type`, `Reason` and `Message` fields based on the

> invocation context.

Reviewer 2: Mention the ones that we have agreed to qualify for `TerminatedByControlPlane`, namely preemption, taint based evictions, eviction API, node cordoning

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

## Alternatives

### Only support for exit codes

We considered supporting just exit codes when defining the policy for handling
pod failures. However, this approach alone would not be sufficient to
distinguish pod failures caused by infrastructure issues. A special handling
of such failures is important in some use cases (see: [Story 2](#story-2)).

### Using Pod status.reason field

We considered using of the pod's `status.reason` field to determine
the reason for a pod failure. This field would be set based on the DeleteOptions
reason field associated with the delete API requests. However, this approach is

> problematic as then the field would be used to set by multiple components

Reviewer 2: Also reasons could be arbitrary strings, making it hard for users to know which reasons to look for in each version.

failures may be encountered late in the execution time of a program.

Sometimes it can be determined from containers exit codes
that the root cause of a failure is in the executable and the
job is destined to fail regardless of the number of retries. However, since
the large workloads are often scheduled to run over night or over the
weekend, there is no human assistance to terminate such a job early.

The need for solving the problem has been emphasized by the kubernetes
community in the issues, see: [#17244](https://github.com/kubernetes/kubernetes/issues/17244) and [#31147](https://github.com/kubernetes/kubernetes/issues/31147).

Some third-party frameworks have implemented retry policies for their pods:
- [TensorFlow Training (TFJob)](https://www.kubeflow.org/docs/components/training/tftraining/)
- [Argo workflow](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml) ([example](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml))

Additionally, some pod failures are not linked with the container execution,
but rather with the internal kubernetes cluster management (see:

> [Scheduling, Preemption and Eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/)).

Reviewer 8: Beyond the errors around Scheduling, Pre-emption and Eviction, there could be numerous other failures not related to container execution that would lead to repeated failures that would require human intervention to recover from. Some examples:
1. Incorrect image specs on containers
2. Incorrect secret/configmap references
3. Incorrect PVC references.
4. Inability of a CSI plugin to attach/mount necessary inline/persistent volumes 
5. General admission failure due to violation of configured policies through OPA/other webhooks.

Are errors like above in scope of the kind of failures that need to be distinguished by the job controller?

Author: In cases 1-3 (which I tested) pods get stuck in the `status.phase=Pending` rather than `Failed`. In case 1 we get a nice dedicated `.status.containerStatuses.waiting.reason=ErrImagePull`, but in cases 2-3 pods get stuck in the `CreatingContainer` status. 

Handling of such errors is out of scope for this KEP. Such configuration issues should be first recognized as Pod failures (not pending pods) which seems deserving a KEP on its own.

> Such pod failures should be recongnized as infrastructure failures and it

Reviewer 9: s/recongnized/recognized/

- Similar termination policies for other workload controllers such as Deployments
  or StatefulSets.

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

## Proposal

Extension of the Job API with a new field which allows to configure the set of
conditions and associated actions which determine how a pod failure is handled.
The extended Job API supports discrimination of pod failures based on the
container exit codes as well as based on the end state of a failed pod.

In order to support discrimination of pod failures based on their end state
we use the already existing `status.conditions` field to append a dedicated Pod
Condition indicating (with its type) that the pod is being terminated by an
internal kubernetes component. Moreover, we modify the internal kubernetes
components to send an API call to append the dedicated Pod condition along with
sending the associated Pod delete request. In particluar, the following
kubernetes components will be modified:
- kube-controller-manager (taint manager performing pod eviction, also: podgc, node-lifecycle-controller)
- kube-scheduler (when performing `Preemption`)

> - kubelet (when performing `NodePressure` eviction)

Reviewer 8: For Kubelet, you may also want to consider the node-level admission failures that lead to a pod ending up in status.phase == Failed (e.g. status.reason set to `OutOfCpu`, `OutOfMemory`, `UnexpectedAdmissionErrorPod`, etc)

Author: We remove kubelet modifications from the KEP (at least for Alpha). The reason being that in many scenarios it is hard to be helpful in recognizing if a pod failure initiated by kubelet should be retried or not. For example, when kubelet initiates node pressure eviction due to memory it may mean either a software bug (in which case we want to terminate the job) or the node being low on memory due to other processes.

Reviewer 2: Admission failures are tricky in a different way. They happen as a result of kube-scheduler and kubelet being out of sync, which can happen when using static pods or when node labels change. So they are kind of a termination by the control plane.
But overall it sounds like there is a lot of debate on what the failures from kubelet are, so we could benefit from an alpha release without accounting for them so we can gather feedback from users.

container restarts within a running pod. However, we consider handling of
job-level failures or termination (for example, a Preemption) as an integral part of
this proposal. Also, when pod's `spec.restartPolicy` is specified as `Never`, then the
failures can't be handled by kubelet and need to be handled at job-level anyway.

Also, we consider this proposal as a natural extension of the already exiting
mechanism for job-level restarts of failed pods based on the Job's
`spec.backoffLimit` configuration. In particular, this proposal aims to fix the
issue, in this mechanism, of unnecessary restarts when a Job can be determined
to fail despite retries.

We believe this feature can co-exist with other pod-level features providing
container restart policies. However, if we establish there is a problematic
interaction of this feature with another such feature, then we will consider
additional validation of the JobSpec configuration to avoid the situation.

If, in the future, we introduce failure handling within the Pod spec, it would
be limited to restartPolicy=OnFailure. Only one of the Pod spec or Job spec APIs
will be allowed to be used at a time.

> #### Limitting this feature

Reviewer 2: ```suggestion
#### Relationship with Pod.spec.restartPolicy
```

const (

	// This is an action which might be taken on pod failure - mark the
	// pod's job as Failed and terminate all running pods.
	BackoffActionTerminate BackoffPolicyAction = "Terminate"

	// This is an action which might be taken on pod failure - the pod will be
	// restarted and the counter for .backoffLimit will not be incremented.
	BackoffActionIgnore BackoffPolicyAction = "Ignore"
)

type BackoffPolicyOnExitCodesOperator string

const (
	BackoffPolicyOnExitCodesOpIn    BackoffPolicyOnExitCodesOperator = "In"
	BackoffPolicyOnExitCodesOpNotIn BackoffPolicyOnExitCodesOperator = "NotIn"
)

type BackoffPolicyOnExitCodesRequirement struct {
	// Restricts the check for exit codes to only apply to the container with
	// the specified name. When empty the rule applies to all containers
	// +optional
	ContainerName *string

	// Represents the relationship between the container exit code(s) and the
	// specified values.
	Operator BackoffPolicyOnExitCodesOperator

	// Specifies the set of values. Each returned container exit code (might be
	// multiple in case of multiple containers) is checked against this set of
	// values with respect to the operator

> 	// +optional

Author: we consider a simple API for Alpha where only specific values are possible to specify. This way, in the first iteration, we don't need to introduce non-standard operators such as InClosedRange etc.

Such pod failures should be recongnized as infrastructure failures and it
should be possible to ignore them from the counter towards `backoffLimit`.

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

### Goals

- Extension of Job API with user-friendly syntax to terminate jobs based on the
  end state of the failed pod/container.

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

### Non-Goals

- Implementation of other indicators of non-retriable jobs such as termination logs.
- Modification of the semantics for job termination. In particular, allowing for
  all indexes of an indexed-job to execute when only one or a few indexes fail
  [#109712](https://github.com/kubernetes/kubernetes/issues/109712).

> - Similar termination policies for other workload controllers such as Deployments

Reviewer 3: I do think it's worth discussing whether this belongs at the job-spec level or the pod-spec level ... something like https://github.com/kubernetes/kubernetes/issues/31147 would require it at the pod-spec level

### Notes/Constraints/Caveats (Optional)

Alternative versions of the API to define requirements on exit codes and on
pod status have been proposed. We discuss them here based on the following
example: let's assume we want to fail a job if the container exit code
(for "main-container") is in range [1,10] or [100,127] . Also, we want
to ignore the pod failure if the pod condition type is "Preempted" or "Evicted".

In the current API we can formulate this policy as follows:

```yaml
 backoffPolicy:
   rules:
   - action: Terminate
     requirements:
     - onExitCodes:
         containerName: main-container
         inRange:
         - minInclusive: 1
           maxInclusive: 10
         - minInclusive: 100
           maxInclusive: 127
   - action: Ignore
     requirements:
     - onPodCondition:
       - key: Status
         operator: In
         values:
         - True
       - key: Type
         operator: In
         values:
         - Preempted
         - Evicted
```

Alternative: Source-based

This approach unifies the API for exit codes and pod conditions using the
concept of an operator:

```yaml
 backoffPolicy:
   rules:

>    - action: Terminate

Reviewer 10: are actions ordered?

Author: The rules (which are containers for a single action and a list of requirements) and ordered and processed from top to bottom.

     requirements:
     - source: ExitCode
       containerName: main-container
       operator: InClosedRange
       values:
       - 1
         10
     - source: ExitCode
       containerName: main-container
       operator: InClosedRange
       values:
       - 100
         127

>    - action: Ignore

Reviewer 10: I like the aspect of being precise in definition - here are exact codes or states that would change the default behavior. And default always stays "retry".

Author: This is true. It is stated in the comment for the backoffPolicy field in JobSpec and emphasized in the Story 1.

Both approaches to the final API make this assumption - if no rule matches the pod failure, then the default handling applies.

  they are not used against maps as the label selector operators are)
- requires introduction of a non-standard operator `InClosedRange` to shorten
  the syntax (compared to using `Lt` and `Gt` operators). The new operator(s)
  can be confusing for users used to label selector operators
- there is a list of conditions (and thus condition types), so the operator
  `In` would be misleading. The operator `AnyIn` would better reflect the
  intention, but it would be another non-standard operator
- it is hard to achieve the flexibility which we can have in the current
  approach. For example, it is not clear how to (1) exclude condition types
  corresponding to conditions with status=False, or how to check presence of
  conditions with Type=Evicted and Reason=NodePressure.

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

#### Evolving condition types

The list of available PodCondition types field will be evolving with new

> values being added and potentially some values becoming obsolete. This can make

Reviewer 10: also test coverage for the different types of failures would become important. I don't think it has a good test coverage today, please add to the test plan

Author: Notes added

> values being added and potentially some values becoming obsolete. This can make

Reviewer 10: it also makes status a contract.  Like this rename:
https://github.com/kubernetes/kubernetes/issues/102820 will be very problematic when we have this kind of dependency.

Reviewer 2: But that's not a problem with ConditionTypes, as they are limited as part of the API https://github.com/kubernetes/api/blob/edebc6706ddef7e828b5b606fbd0c91c6843c2fa/core/v1/types.go#L2646-L2656

I think this is another reason not to filter by `reason` :)

Author: Agree, for that matter, we should encourage only the use of `type` in the API and discourage the use of `reason`

condition with dedicated `Type`, `Reason` and `Message` fields based on the
invocation context.

### JobSpec API

We extend the Job API in order to allow to apply different actions depending
on the conditions associated with the pod failure.

```golang
// BackoffPolicyAction specifies how a Pod failure is handled.
type BackoffPolicyAction string

const (

	// This is an action which might be taken on pod failure - mark the
	// pod's job as Failed and terminate all running pods.
	BackoffActionTerminate BackoffPolicyAction = "Terminate"

	// This is an action which might be taken on pod failure - the pod will be
	// restarted and the counter for .backoffLimit will not be incremented.
	BackoffActionIgnore BackoffPolicyAction = "Ignore"
)

type BackoffPolicyOnExitCodeInterval struct {
	MinInclusive int
	MaxInclusive int
}

type BackoffPolicyOnExitCodeRequirement struct {
	// Restricts the check for exit codes to only apply to the container with
	// the specified name. When empty the rule applies to all containers
	// +optional
	ContainerName *string

	// Specifies the set of intervals. The requirement is satisfied when the
	// actual exit code is in any of the intervals
	// +optional

> 	InRange []BackoffPolicyOnExitCodeInterval

Reviewer 2: I prefer we have an operator field and a single `values []int` to be closer to node affinity.

Author: I'm happy with going this path if there is such consensus.

I was a little bit concern that by making it closer to (but not exactly like) the node affinity we can make it confusing. The syntax would not make use of the "key" field as there. Also, a set of non-standard operators like InClosedRange.

[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/

Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
-->
#### Alpha

- Implementation:
  - handling of failed pods with respect to `spec.backoffPolicy` by Job controller
  - appending of a dedicated Pod condition (when the Pod termination is
    initiated by a kubernetes control plane component) to the list of Pod
    conditions along with sending the Pod delete request
  - define as a constant and document the new Pod condition Type
  - the feature is limited by disallowing of the use of `onExitCodes` when
    `restartPolicy=OnFailure`
- The feature flag disabled by default
- Tests: unit and integration

#### Beta

- Address reviews and bug reports from Alpha users
- E2e tests are in Testgrid and linked in KEP
- A scalability test to demonstrate the limited impact of the additional API call
  when terminating a Pod
- Re-evaluate modification to kubelet to send a dedicated condition when

>   terminating a Pod

Reviewer 2: ```suggestion
  terminating a Pod, based on user feedback.
```

Maybe we can be a bit more intentional here. We hope to clarify from user feedback, which can be obtained by creating a form and bringing the question to wg-batch and sig-node

-->

- [ ] Events
  - Event Reason:
- [ ] API .status
  - Condition name:
  - Other field:
- [ ] Other (treat as last resort)
  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

<!--
This is your opportunity to define what "normal" quality of service looks like
for a feature.

It's impossible to provide comprehensive guidance, but at the very
high level (needs more precise definitions) those may be things like:
  - per-day percentage of API calls finishing with 5XX errors <= 1%
  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99.9% of /health requests per day finish with 200 code

These goals will help you determine what you need to measure (SLIs) in the next
question.
-->

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

<!--
Pick one more of these and delete the rest.
-->

- [x] Metrics
  - Metric name:
    - `job_sync_duration_seconds` (existing): can be used to see how much the
feature enablement increases the time spent in the sync job

>     - `job_finished_total` (existing, extended by a label): the new `reason`

Reviewer 2: The metrics below are not so much about SLI, but they can answer the earlier question:
How can an operator determine if the feature is in use by workloads?

      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?

> -->

Reviewer 2: Maybe mention a Pod level API here with regard to exit code.

Author: Possibly, but it does not feel to me like a drawback. I keep the discussion on the pod-level vs. job-level decision in Notes.

For Alpha, we limit this feature by not recognizing Pod failures initiated
by Kubelet. This is because it is hard to determine in some scenarios of
Pod failures initiated by Kubelet if they should be retried or should not.
For example, when Kubelet evicts a pod due to node pressure it might mean either
software bug (then it might be better to terminate the entire job) or the node
being low on memory due to other processes (in which case it is sensible to retry).
We are going to re-evaluate for Beta if we want to add support for recognizing
pod terminations initiated by Kubelet.

#### Current state review

Here we review the current state of kubernetes (version 1.24) regarding its
handling pod failures in different scenarios we tested it. In particular,
we list the most important pod fields which could be used as indicators if
the pod failure should or should not be retried.

The examples below demonstrate that there is no universal indicator (like a
pod or container field) currently that discriminates pod failures which should
be retried from those which should not be retried.

Hence, we propose to introduce a new dedicated Pod condition type appended

> in case of Preemption and Taint-based eviction.

Reviewer 2: If this is the proposal, it needs to be in the "Design Details".

- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
  - [Only support for exit codes](#only-support-for-exit-codes)
  - [Using Pod status.reason field](#using-pod-statusreason-field)
  - [Using of various PodCondition types](#using-of-various-podcondition-types)
  - [More nodeAffinity-like JobSpec API](#more-nodeaffinity-like-jobspec-api)
- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
<!-- /toc -->

## Release Signoff Checklist

<!--
**ACTION REQUIRED:** In order to merge code into a release, there must be an
issue in [kubernetes/enhancements] referencing this KEP and targeting a release
milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
of the targeted release**.

For enhancements that make changes to code or processes/procedures in core
Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
Signoff checklist to be completed.

Check these off as they are completed for the Release Team to track. These
checklist items _must_ be updated for the enhancement to be released.
-->

Items marked with (R) are required *prior to targeting to a milestone / release*.

> - [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)

Reviewer: Nit: some of these should be checked.

  - `state=Ternminated`
  - `exitCode=137`
  - `reason=Error`
- Retriable: Unclear, excessive disk usage suggests a bug or misconfiguration.
  However, a restart on another node may succeed

##### OOM kill

- Reproduction: We run a job with a pod which attempts to allocate more
  memory than constrained in the container spec by `resources.limits.memory`
- Comments: handled by kubelet
- Pod status:
  - status: OOMKilled
  - `phase=Failed`
  - `reason=`
  - `message=`
- Container status:
  - `state=Ternminated`
  - `exitCode=137`
  - `reason=OOMKilled`
- Retriable: Unclear, but if occurs when
  `resources.requests.memory=resources.limits.memory` it strongly suggests
  a software bug.

##### Node drain

- Reproduction: We run a job with a long-running pod, then drain the node
  with the `kubectl drain` command
- Comments: performed by Eviction API
- Pod status:
  - status: Terminating
  - `phase=Failed`
  - `reason=`
  - `message=`
- Container status:
  - `state=Ternminated`
  - `exitCode=137`
  - `reason=Error`
- Retriable: Yes

##### Disconnected node

- Reproduction: We run a job with a long-running pod, then disconnect the node
  and delete it by the `kubectl delete` command

> - Comments: unclear handling, the pod remains in the `Running` phase. Potentially,

Author: Or investigate why it is not moved to Failed phase and "fix" - but it seems outside the KEP. 

Alternatively, just leave this scenario as outside the scope of the KEP and focus on those which are in Failed phase
only, because I think matching running phase could be complicated in general.

	BackoffPolicyOnExitCodesOpNotIn BackoffPolicyOnExitCodesOperator = "NotIn"
)

type BackoffPolicyOnExitCodesRequirement struct {
	// Restricts the check for exit codes to only apply to the container with
	// the specified name. When empty the rule applies to all containers
	// +optional
	ContainerName *string

	// Represents the relationship between the container exit code(s) and the
	// specified values.
	Operator BackoffPolicyOnExitCodesOperator

	// Specifies the set of values. Each returned container exit code (might be
	// multiple in case of multiple containers) is checked against this set of
	// values with respect to the operator
	Values []int
}

type BackoffPolicyOnPodConditionsOperator string

const (
	BackoffPolicyOnPodConditionsOpIn BackoffPolicyOnPodConditionsOperator = "In"
)

type BackoffPolicyOnPodConditionsRequirement struct {

	// Represents the relationship between the actual Pod condition types
	// and the set of specified Pod condition types
	Operator BackoffPolicyOnPodConditionsOperator

	// Specifies the set of values. Each actual pod condition type is checked
	// against this set of values with respect to the operator

> 	Values []api.PodConditionType

Reviewer 2: I would name this types, for clarity. Also you need to say in the documentation that it only matches conditions with Status=True

Author: Added a note on the status=True.

I like Values as it is more similar to the "onExitCodes" API. Maybe instead we say "onPodConditionTypes" level up?

Reviewer 2: that might work

Reviewer 2: We can leave this kind of low level details to code review.

  backoffLimit: 3
  backoffPolicy:
    rules:
    - action: Terminate
      onContainerExitCode:
        notIn:
          min: 40

>           max: 50

Author: I think optional max could be misleading as it can be interpreted as interval which is only bounded from left. 
But I get the point about API to express a single exit code easy - let me think and come up with something

>           max: 50

Author: If someone wants to target a single exit code (say 42), it can be spelled in the new API:

```
    - action: Terminate
      requirements:
      - source: ExitCode
        operator: In
        Values:
        - 42
```

>           max: 50

Author: Update, in the current version of the API it is possible to express this as follows:

```
   - action: Terminate
     requirements:
     - onExitCodes:
        in:
        - 42
```

together with explanation why this is acceptable.
-->

Unit tests will be added along with any new code introduced. In particular,
the following scenarios will be covered with unit tests:
- handling of `spec.backoffPolicy` by the Job controller when the feature gate is
  enabled,
- ignoring of `spec.backoffPolicy` by the Job controller when the feature gate is

>   disabled,

Reviewer 2: kube-apiserver also handles validating Job (in particular `.spec.backoffPolicy`) and writing it to etcd.

### Version Skew Strategy

> The API exposed by kube-apiserver is extended by an optional field `reason` in

Reviewer 2: Your current working is not talking about version skew.

For this feature, there is a version skew when upgrading the control plane: one kube-apiserver could be in version N, while kubes-scheduler and kube-controller-manager are in N-1. https://kubernetes.io/releases/version-skew-policy/#kube-controller-manager-kube-scheduler-and-cloud-controller-manager

The strategy we are using is the feature gate. In version N, the feature gate is disabled, so nobody is writing the reason. In version N+1, the feature gate is enabled, but all the kube-apiservers at version N know how to handle the field.

Author: Section updated after the switch to use conditions

> ## Alternatives

Reviewer 2: Great summary!
